resolved stochastic rounding flow for fp32->bf8

nkmellem opened the pull request.

alheinecke commented on src/libxsmm_lpflt_quant.c:
```diff
@@ -724,22 +724,55 @@ LIBXSMM_API void libxsmm_convert_bf8_f32(const libxsmm_bfloat8* in, float* out,
 
-LIBXSMM_API void libxsmm_stochastic_convert_fp32_bf8(const float* in, libxsmm_bfloat8* out, unsigned int len) {
+LIBXSMM_API void lsfr_Xwide( unsigned int* rng_state, float* prng_out, const unsigned int width ) {
+  union { unsigned int i; float f; } rng_num = { 0 };
```
why float out? we are going to use it later as 8bit values, and int32 should be returned

nkmellem replied:
Fixed

alheinecke commented on src/libxsmm_lpflt_quant.c:
```diff
@@ -724,22 +724,55 @@ LIBXSMM_API void libxsmm_convert_bf8_f32(const libxsmm_bfloat8* in, float* out,
 }
 
-LIBXSMM_API void libxsmm_stochastic_convert_fp32_bf8(const float* in, libxsmm_bfloat8* out, unsigned int len) {
+LIBXSMM_API void lsfr_Xwide( unsigned int* rng_state, float* prng_out, const unsigned int width ) {
```
Why is this under a public API and not LIBXSMM_API_INTERN. I feel it should be LIBXSMM_API_INTERN.
Also the function is not defined in the header.
As C has no namespaces, please prepend libxsmm_

alheinecke replied:
why is the routine "vectorized", when used scalar further below?

nkmellem replied:
fixed it

alheinecke commented on src/libxsmm_lpflt_quant.c:
```diff
@@ -724,22 +724,55 @@ LIBXSMM_API void libxsmm_convert_bf8_f32(const libxsmm_bfloat8* in, float* out,
     /* perform round nearest tie even */
     if ( do_round != 0) {
-      if ( (short_round & 0x7c00) == 0x0000) {
+      if ( (short_round & 0x7c00) == 0x0000 && (short_round & 0x7fff) > 0x0000 ) {
```
why this condition? it's always true except -0? But then the first condition cannot be true and they are anded.

nkmellem replied:
removed it.

alheinecke commented on src/libxsmm_lpflt_quant.c:
```diff
@@ -724,22 +724,55 @@ LIBXSMM_API void libxsmm_convert_bf8_f32(const libxsmm_bfloat8* in, float* out,
-
+    float vrng[16];
+    lsfr_Xwide( (unsigned int*)rng_state, vrng, 1 );
+    unsigned short rand = (unsigned short) vrng[1];
```
why taking the 2nd and not the lowest number? (still see above why this is a vectorize prng for a scalar function?

alheinecke commented on src/generator_common_x86.c:
```diff
@@ -3033,7 +3037,7 @@ void libxsmm_generator_vcvtneps2bf8_avx512_prep_stack ( libxsmm_generated_code*
 
   /* push 0x1 on the stack, fixup */
-  libxsmm_x86_instruction_alu_imm( io_generated_code, LIBXSMM_X86_INSTR_MOVQ, io_gp_reg, 0x00010001);
+  libxsmm_x86_instruction_alu_imm( io_generated_code, LIBXSMM_X86_INSTR_MOVQ, io_gp_reg, 0x01000100);
```
the RNE Fix needs to be on the 16bit type. why is this now on the 8bit type?

nkmellem replied:
it is still 16-bit value , shifted by 8 bits 0x0100 instead if 0x0001

alheinecke reviewed (changes requested):
Please address all the comments/questions raised.

Additionally, before we can merge this, unit tests needs to be enabled/added. Minimally for the COPY TPP and ADD TPP.
https://github.com/libxsmm/libxsmm/blob/main/samples/eltwise/kernel_test/generate_unary_simple_test_scripts.sh
https://github.com/libxsmm/libxsmm/blob/main/samples/eltwise/kernel_test/generate_binary_test_scripts.sh

And allow -list for non AVX512 and non-Intel platforms need to be updated:
e.g. for AVX2/HSW: 
https://github.com/libxsmm/libxsmm/blob/main/samples/eltwise/run_test_hsw.txt

alheinecke reviewed (changes requested):
Can you please rebase to the branch I had created... we have now changes coming in which should not e there...

nkmellem commented:
Addressed all the issues.

alheinecke commented on samples/eltwise/kernel_test/generate_binary_test_scripts.sh:
```diff
@@ -11,46 +11,61 @@ trap 'rm ${TMPFILE}' EXIT
+
+        PREC_OUT=$(echo "$PRECLC" |  awk -F"_" '{print $2}')
+        if [ "$PREC_OUT" == 'bf8' ] ; then
+          PREC_IN=$(echo "$PRECLC" |  awk -F"_" '{print $1}')
```
why is append just "stoch", etc. not enough?
please rework that all combos with PREC_OUT and "stochastic" use continue instead of writing the same file to disk as for non-stochastic.

nkmellem replied:
modified accordingly

alheinecke commented on samples/eltwise/kernel_test/generate_unary_simple_test_scripts.sh:
```diff
@@ -11,73 +11,89 @@ trap 'rm ${TMPFILE}' EXIT
-      OUTNAME=${OUTNAME}${TPPNAME}_${PRECLC}_${LD}.sh
+        PREC_OUT=$(echo "$PRECLC" |  awk -F"_" '{print $2}')
+        if [ "$PREC_OUT" == 'bf8' ] ; then
+          PREC_IN=$(echo "$PRECLC" |  awk -F"_" '{print $1}')
```
see comments in binary

nkmellem replied:
done

alheinecke commented on samples/eltwise/eltwise_unary_simple.c:
```diff
@@ -42,7 +42,7 @@
 #define EXP_OP 17
 #define REPLICATE_COL_VAR 27
-#if 0
+#if 10
```
please use either #if 0 or #if 1 or #ifdef xyz

alheinecke commented on src/generator_common_x86.c:
```diff
@@ -2206,6 +2206,56 @@ void libxsmm_generator_xoshiro128p_f32_avx2_avx512( libxsmm_generated_code* io_g
 }
 
+LIBXSMM_API_INTERN
+void libxsmm_generator_xoshiro128p_i32_avx2_avx512( libxsmm_generated_code* io_generated_code,
```
(also definition in header is missing, in case this routine is needed)

nkmellem replied:
done

alheinecke reviewed (changes requested):
Some minor issues are still open.

Also the  allow list in samplex/eltwise/run_text_[arch].txt are not updated.

there are also conflicts with target (fix_bf8_stoch_round) branch & I see definition of unused variables. The first 1 needs to be resolved before merging is possible.

nkmellem commented on src/libxsmm_lpflt_quant.c:
```diff
@@ -724,30 +724,87 @@ LIBXSMM_API void libxsmm_convert_bf8_f32(const libxsmm_bfloat8* in, float* out,
 
-LIBXSMM_API void libxsmm_stochastic_convert_fp32_bf8(const float* in, libxsmm_bfloat8* out, unsigned int len) {
+LIBXSMM_API_INTERN void libxsmm_lsfr_i32( unsigned int* rng_state, unsigned int* prng_out, const unsigned int seed_idx ) {
+  libxsmm_float_uint rng_num;
```
fixed it

nkmellem commented on src/libxsmm_lpflt_quant.c:
```diff
@@ -724,30 +724,87 @@ LIBXSMM_API void libxsmm_convert_bf8_f32(const libxsmm_bfloat8* in, float* out,
 }
 
-LIBXSMM_API void libxsmm_stochastic_convert_fp32_bf8(const float* in, libxsmm_bfloat8* out, unsigned int len) {
+LIBXSMM_API_INTERN void libxsmm_lsfr_i32( unsigned int* rng_state, unsigned int* prng_out, const unsigned int seed_idx ) {
```
done

alheinecke commented:
what is the issue with adding a **/pp scalar version and then we use the correct LSFR?

nkmellem commented:
> what is the issue with adding a **/pp scalar version and then we use the correct LSFR?

From the code, it seems this is not any standard validated implementation. Is there reference version of this implementation that reliably produces a uniform distribution? if it exists, we can switch to that.

nkmellem commented:
> https://prng.di.unimi.it/xoshiro128plusplus.c (as the name suggests). Where is the the mismatch with that code?

Updated the reference code. removed the new _i32 code gen version

hfp commented:
Further commits to this branch (prior to merge) shall first merge/rebase main. This is because CI changed some recipes now requiring updated scripts as only present in main branch. If no further work is needed, it can be merged right away.

alheinecke commented:
Thanks @hfp; main merged.
@nkmellem : I assume you are still working on the allow-lists?

alheinecke reviewed (commented):
LGTM, waiting for the remaining allow-lists

nkmellem commented:
> Thanks @hfp; main merged. @nkmellem : I assume you are still working on the allow-lists?

I have done the HSW list. I was not planning to do the others. Its manual grunt work of copy-pasting. Is there a smarter way to generate these lists?

alheinecke commented:
Nope
